### DISCUSSION TITLE
Move away from Electron?

### DISCUSSION THREAD

**Discussion by zerebos** [TriliumNext/Trilium#6077] [General]
I think we're all pretty much on the same page to stick with Electron as it's been working well and the footprint is still fairly low compared to other Electron apps.

But as I've been thinking about the massive potential changes we are discussing I think this should be given a bit more thought. Using Electron limits us a bit in that we are really a desktop and web only application. Putting in a lot of work to potentially overhaul the UI, convert to Svelte and TypeScript it would be a shame to not be able to deliver a mobile experience.

There are some existing systems that allow a web frontend that can deploy to mobile. The most prolific that comes to mind is [Adobe Cordova](https://cordova.apache.org/). If you've ever heard of the password manager [Padloc](https://padloc.app/), this is what they currently use to do web, desktop, mobile, and browser extension. The other one is one previously mentioned in other threads, [Tauri](https://tauri.app/). Technically Tauri's mobile support hasn't fully released as stable but the beta is just getting better and better, it won't be too terribly long until it is released.

From my understanding, Tauri is more similar to Electron in terms of the development process. The "only" change would be rewriting the entire backend to be in Rust instead of JS/TS. As for Cordova, it seems to be more of a framework and would probably require even more work.

That said while it would be a massive change to move away from Electron, there's also potentially the benefit of supporting mobile with the same codebase while also lowering our memory footprint and speeding up the backend.

Looking forward to seeing everyone's thoughts.

**Comment by eliandoran** [comment]
@rauenzi , for me the Electron app has been quite OK, so I would not feel an immediate need to move away from it. I would prefer, at least for now, to not base ourselves on webview wrappers due to potential issues with the possible outdated native browsers. However, if I come to think about it, we already have to support multiple browsers because you can access the web interface of the server from any browser.

Mobile support however is a must, and I don't mean just a PWA, but a full-fleged Trilium instance. Honestly, this has been quite a deal-breaker for me especially in using Trilium for notes. I always ended up having to write in my notes app and then having to move the notes back into Trilium once at the PC.

I have used Cordova before and it felt quite decent, although I used it in situations where the server had little to no logic in it. The more challenging part would be to embed the current server in it, but I'm pretty sure we can manage.

**Reply by zerebos** [reply]
Yeah mobile support is the big killer for me personally as a user as well. It's honestly the main reason I even bring up potentially moving off of Electron. I don't know any way to bring an Electron app to mobile.

> However, if I come to think about it, we already have to support multiple browsers because you can access the web interface of the server from any browser.

This was exactly my thought on this. If we are already adding support for general browsers, webviews shouldn't be a problem. Modern webviews are almost all chromium based anyways.

I'm glad you know more about Cordova as I know so little, but I do know about Tauri!

**Comment by zadam** [comment]
I think "mobile support" has two related, but distinct meanings:

1) being able to run the whole Trilium on a phone (discussed here), offline support, synchronization support etc.

2) having a good mobile UI. I believe that many users would be actually quite happy with a nice phone UI, even if it was online-only, connecting to a server. But this is also a lot of work, Trilium's features have been designed for the desktop, some of them are "too complex to fit onto a phone screen", or at least they have to be presented in a different way (e.g. through a wizard). Maybe a new frontend can be designed in such a (responsive) way that it works fine on both desktop and mobile.

**Reply by zerebos** [reply]
I agree those are two different ideas overall, app vs pwa. But I think based on the options we discussed above (Tauri + Cordova) to tackle point 1, the app frontend itself would still be web based so your number 2 would still be a requirement for us in some way or form. That said, like you mentioned in another thread, it's not a good idea to tackle too many ideas at once, so switching to TypeScript plus a standardized frontend framework would allow us to immediately tackle point 2 while we consider point 1.

**Comment by shutaozhenzhen** [comment]
I don't think trilium has PWA function yet?
I think add PWA function to trilium is useful.

**Reply by zerebos** [reply]
You can install Trilium as a PWA currently, it's just always-online.

**Comment by Dialga** [comment]
What about porting to [webui](https://github.com/webui-dev/webui)? There is also a [deno port](https://github.com/webui-dev/deno-webui).

**Comment by capi** [comment]
I personally use it as PWA against a locally running docker container (actually podman), which then syncs with a remote instance. Works identically well for me like the Electron app, without needing to have another Chrome instance running. (PWA is used in Chrome and Edge in my case). I personally think it would be best to keep browser/PWA mode a first-class citizen since it's the most portable one and concentrate on how to deploy the backend side wherever it is needed.
